sympy/logic/algorithms/lra_theory.py

```diff
@@ -0,0 +1,774 @@
+from sympy.solvers.solveset import linear_eq_to_matrix
+from sympy.matrices.dense import eye
+from sympy.assumptions import Predicate
+from sympy.assumptions.assume import AppliedPredicate
+from sympy.assumptions.ask import Q
+from sympy.core import Dummy
+from sympy.core.mul import Mul
+from sympy.core.add import Add
+from sympy.core.relational import Eq, Ne
+from sympy.core.sympify import sympify
+from sympy.core.singleton import S
+from sympy import SYMPY_DEBUG
+from sympy.core.numbers import Rational, oo
+from sympy.matrices.dense import Matrix
+from sympy.matrices.common import MatrixKind
+from sympy.core.kind import NumberKind
+
+class UnhandledNumber(Exception):
+    pass
+
+ALLOWED_PRED = {Q.eq, Q.gt, Q.lt, Q.le, Q.ge}
+HANDLE_NEGATION = True
+
+class LRASolver():
+    """
+    Linear Arithmetic Solver for DPLL(T) implemented with an algorithm based on
+    the Dual Simplex method. Uses Bland's pivoting rule to avoid cycling.
+
+    TODO:
+     - Implement and utilize backtracking
+     - Handle non-rational real numbers
+     - Handle positive and negative infinity
+     - Simplify matrix by removing unused variables using Gaussian elimination
+
+    References
+    ==========
+
+    .. [1] Dutertre, B., de Moura, L.:
+           A Fast Linear-Arithmetic Solver for DPLL(T)
+           https://link.springer.com/chapter/10.1007/11817963_11
+    """
+
+    def __init__(self, A, slack_variables, nonslack_variables, enc_to_boundary, x_subs, s_subs, testing_mode):
+        """
+        Use the "from_encoded_cnf" method to create a new LRASolver.
+        """
+        self.run_checks = testing_mode
+        self.x_subs = x_subs
+        self.s_subs = s_subs
+
+        if any(not isinstance(a, Rational) for a in A):
+            raise UnhandledNumber
+        if any(not isinstance(b.bound, Rational) and (b.bound != oo) and (b.bound != -oo) for b in enc_to_boundary.values()):
+            raise UnhandledNumber
+        m, n = len(slack_variables), len(slack_variables)+len(nonslack_variables)
+        if m != 0:
+            assert A.shape == (m, n)
+
+        if self.run_checks:
+            assert A[:, n-m:] == -eye(m)
+
+        self.enc_to_boundary = enc_to_boundary
+        self.boundary_to_enc = {value: key for key, value in enc_to_boundary.items()}
+        self.A = A
+        self.slack = slack_variables
+        self.nonslack = nonslack_variables
+        self.all_var = nonslack_variables + slack_variables
+
+        self.slack_set = set(slack_variables)
+
+        self.is_sat = True
+        self.result = None  # always one of: (True, assignment), (False, conflict clause), None
+
+    @staticmethod
```

Comment:
This should be a classmethod.

Reply:
I have thought similar that staticmethod can always be redundant in Python before, especially for everything was duck typed.

However, now I just think that there is also no use of classmethod if the method does not depend on the parameter cls.

Formally, it should be possible to substitute  staticmethod, with ‘def function’, as well as trivial class methods or trivial bound methods that does not depend on the class, object.

So I stopped thinking that that coding was a problem, especially after type checkers or linters are established.

Reply:
The reason I say it should be a class method is because it's an alternative constructor, which is the typical use-case of class methods in Python. However, in this case, it returns an instance of the class and also something else, so maybe it isn't really a constructor.

Reply:
Also this would depend on cls. The function uses `LRASolver`, which would be `cls` if this were a class method.

Reply:
Okay, I thought it was pointing out to `pivot` though